src/transaction/log_page_buffer.c

```diff
@@ -3361,6 +3364,250 @@ logpb_write_toflush_pages_to_archive (THREAD_ENTRY * thread_p)
+      log_Pb.partial_append.status = LOGPB_APPENDREC_IN_PROGRESS;
+      break;
+    case LOGPB_APPENDREC_IN_PROGRESS:
+    case LOGPB_APPENDREC_PARTIAL_FLUSHED_END_OF_LOG:
+      /* keep the status */
+      break;
+    default:
+      /* invalid state */
+      assert_release (false);
+    }
+}
+
+/*
+ * logpb_catchup_end_append - Finish appending a page during catch-up.
+ *      If the appended page has only parital record in it, keep partial appending status.
+ *      Otherwise, complete appending pages, in the partial appending case, or a page.
+ *        - Set the append_lsa and prev_lsa to valid ones.
+ *        - Remove the temporary EOF If log pages are flushed halfway through.
+ *        - Set the partial appending status to LOGPB_APPENDREC_SUCCESS.
+ * 
+ *    See logpb_catchup_start_append() for details of the status transition.
+ *
+ *   log_pgptr(in): a log page pointer appended
+ */
+static void
+logpb_catchup_end_append (THREAD_ENTRY * thread_p, const LOG_PAGE * const log_pgptr)
+{
+  assert (LOG_CS_OWN_WRITE_MODE (thread_p));
+
+  const bool have_only_partial_record_in_page = log_pgptr->hdr.offset == NULL_LOG_OFFSET;
+
+  if (have_only_partial_record_in_page)
+    {
+      // Do nothing. It will be finished up at a following call.
+      return;
+    }
+
+  assert (log_Gl.hdr.append_lsa.pageid == log_pgptr->hdr.logical_pageid);
+  log_Gl.hdr.append_lsa.offset = log_pgptr->hdr.offset;
+
+  LOG_RECORD_HEADER *log_rec = LOG_GET_LOG_RECORD_HEADER (log_pgptr, &log_Gl.hdr.append_lsa);
+  log_Gl.append.prev_lsa = log_rec->back_lsa;
+
+  switch (log_Pb.partial_append.status)
+    {
+    case LOGPB_APPENDREC_IN_PROGRESS:
+      /* success, fall through */
+      break;
+    case LOGPB_APPENDREC_PARTIAL_FLUSHED_END_OF_LOG:
+      {
+	LOG_RECORD_HEADER *origin_append_log_record = LOG_GET_LOG_RECORD_HEADER (log_pgptr, &log_Gl.hdr.append_lsa);
+
+	// The temporary EOF log record at the prev_lsa has to be removed.
+	log_Pb.partial_append.status = LOGPB_APPENDREC_PARTIAL_ENDED;
+	logpb_flush_all_append_pages (thread_p);
+	assert (log_Pb.partial_append.status == LOGPB_APPENDREC_PARTIAL_FLUSHED_ORIGINAL);
+
+	// A new EOF has been appended at the append_lsa in logpb_flush_all_append_pages. Remove it as well.
+	LOG_RECORD_HEADER *append_log_rec = LOG_GET_LOG_RECORD_HEADER (log_Gl.append.log_pgptr, &log_Gl.hdr.append_lsa);
+	assert (append_log_rec->type == LOG_END_OF_LOG);
+	assert (log_Gl.append.log_pgptr->hdr.logical_pageid == log_pgptr->hdr.logical_pageid);
+	memcpy (append_log_rec, origin_append_log_record, sizeof (LOG_RECORD_HEADER));
+
+	logpb_set_dirty (thread_p, log_Gl.append.log_pgptr);
+      }
+      break;
+    default:
+      /* invalid state */
+      assert_release (false);
+    }
+  log_Pb.partial_append.status = LOGPB_APPENDREC_SUCCESS;
+}
+
+/*
+ * logpb_catchup_finish - finish the catch-up job. 
+ *      This sets the prev_lsa and append_lsa to the latest ones traversing log pages from known log record.
```

Comment:
Unfortunately I don't fully understand what this function does. Could you expand the comment a bit please?

From what I understood, it traverses the log records of either:
- the last added log page started from the first complete log record in that page
- or the last log record when that last log record is a big one spanning multiple log pages

and
 - retains the LSA of the last log record as the `log_Gl.append.prev_lsa`
 - while `log_Gl.hdr.append_lsa`/`catchup_lsa` is the LSA of the next log record (still not present) to be added

Is this correct?

I think you could have avoided this extra function with the price of passing a `log_catchup_context` pointer struct to the `logpb_catchup_append_page` and `logpb_catchup_end_append` functions and extracting the value from there.

Not necessary to change the code in this direction, but I am wondering if it is possible.

Reply:
> the last added log page started from the first complete log record in that page
or the last log record when that last log record is a big one spanning multiple log pages

Correct.

> retains the LSA of the last log record as the log_Gl.append.prev_lsa
while log_Gl.hdr.append_lsa/catchup_lsa is the LSA of the next log record (still not present) to be added

I don't understand your words. Probably, we have a misunderstanding. Let me clarify what this function does (my intention).
After appending a page, we don't have precise append_lsa and prev_lsa. Our goal here is to make them have appropriate ones. Basically, they have values as the following which are set in `logpb_catchup_end_append()`:
- append_lsa: log_pgptr->hdr.offset, which is the location we can guarantee there is a log record. 
- prev_lsa: the prev_lsa of the log record append_lsa points to.

This is what has to be kept while appending pages, but it's not what they really mean. The `prev_lsa` is supposed to point to the last log record and the `append_lsa` is supposed to point to the next one, which has not been appended yet. So, here at the end of the whole catch-up process, they are set as they're supposed to be.

How: we start from the most advanced log record we can know, which the `prev_lsa` points to. The `append_lsa` is not reliable because if the last page has no `log_pgptr->hdr.offset`, it just points to the start of the page. And we just traverse log records until we met the last log record.

We can make them every time after appending a page, but we don't need to. So, I do it at the end only once.

Reply:
> The prev_lsa is supposed to point to the last log record and the append_lsa is supposed to point to the next one, which has not been appended yet. So, here at the end of the whole catch-up process, they are set as they're supposed to be.

Ok. I knew [at least a part of] that.

> We can make them every time after appending a page, but we don't need to. So, I do it at the end only once.

Yes, it also crossed my mind that my suggestion to use a `log_catchup_context` pointer ends up with doing this investigation many times for each log page added.